plugin/hpsa/hpsa.py

```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+            cmd2 = ['ctrl', 'slot=%s' % ctrl_num, 'modify', 'nbwc=disable']
+        elif wcp == Volume.WRITE_CACHE_POLICY_WRITE_THROUGH:
+            cmd1.extend(['modify', 'aa=disable'])
+        else:
+            raise LsmError(ErrorNumber.PLUGIN_BUG,
+                           "Got unknown wcp: %d" % wcp)
+
+        try:
+            self._sacli_exec(cmd1, flag_force=True, flag_convert=False)
+            if cmd2:
+                self._sacli_exec(cmd2, flag_force=True, flag_convert=False)
+        except ExecError as exec_error:
+            # Check whether we got SSD volume
+            if self._is_ssd_volume(volume):
+                raise LsmError(
+                    ErrorNumber.NO_SUPPORT,
+                    "HP SmartArray does not allow changing SSD volume's "
+                    "cache policy while SmartPath is enabled")
+            raise exec_error
+
+    @_handle_errors
+    def volume_read_cache_policy_set(self, volume, rcp, flags=Client.FLAG_RSVD):
+        """
+        Depending on this command:
+            To disable both read and write cache:
+                hpssacli ctrl slot=3 ld 1 modify aa=disable
+        """
+        (ctrl_num, array_num, ld_num) = self._cal_of_lsm_vol(volume)
+
+        cmd = ['ctrl', 'slot=%s' % ctrl_num, 'ld', ld_num, 'modify']
+        if rcp == Volume.READ_CACHE_POLICY_DISABLED:
+            cmd.append('aa=disable')
+        elif rcp == Volume.READ_CACHE_POLICY_ENABLED:
+            cmd.append('aa=enable')
+        else:
+            raise LsmError(ErrorNumber.PLUGIN_BUG,
+                           "Got unknown rcp: %d" % rcp)
+        try:
+            self._sacli_exec(cmd, flag_force=True, flag_convert=False)
+        except ExecError as exec_error:
+            # Check whether we got SSD volume
+            if self._is_ssd_volume(volume):
+                raise LsmError(
+                    ErrorNumber.NO_SUPPORT,
+                    "HP SmartArray does not allow changing SSD volume's "
+                    "cache policy")
+            raise exec_error
+
+    def _cal_of_lsm_vol(self, lsm_vol):
+        """
+        Retrieve controller slot number, array number, logical disk number
+        of given volume. Also validate the existence.
+        Return (ctrl_num, array_num, ld_num)
+        """
+        try:
+            new_lsm_vol = \
+                self.volumes(search_key='id', search_value=lsm_vol.id)[0]
+        except IndexError:
+            raise LsmError(
+                ErrorNumber.NOT_FOUND_VOLUME,
+                "Volume not found")
+
+        return new_lsm_vol.plugin_data.split(":")
+
+    def _is_ssd_volume(self, volume):
+        pool = self.pools(search_key='id', search_value=volume.pool_id)[0]
+        disk_ids = self.pool_member_info(pool)[2]
+        lsm_disks = self.disks()
+        for lsm_disk in lsm_disks:
+            if lsm_disk.disk_type == Disk.TYPE_SSD and \
+               lsm_disk.id in disk_ids:
+                return True
```

Comment:
Wouldn't it be better to iterate through the disks that make up the pool instead of iterating though all the disks and then checking if the disk is in the pool?